compiler/tests/test_identify_token.cpp

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright: Copyright nsoul97 (Soulounias Nikolaos) 2019.
+ * License:   $(LINK2 http://www.boost.org/LICENSE_1_0.txt, Boost License 1.0)
+ * Authors:   nsoul97 (Soulounias Nikolaos)
+ */
+
+#include "../identify_token.h"
+#include "../token_name.h"
+#include "../tokens.h"
+#include <cassert>
+#include <cstdio>
+
+extern Token token;
+
+/*
+If the global variable token has its kind member set to the expected_enum, the
+function prints a success message. Otherwise, the program execution stops due to
+the assertion.
+*/
+static void test(const char *input, TOK expected_enum) {
+  identify_token(input);
+  assert(token.kind == expected_enum);
+  if (expected_enum != TOK::UNDEFINED)
+    printf("SUCCESS: function identify_token(\"%s\") sets token's kind member "
+           "to %s token.\n",
+           input, token_name(expected_enum));
+  else
+    printf(
+        "SUCCESS: function identify_token(\"%s\") sets token's kind member to "
+        "UNDEFINED token.\n\t \"%s\" starts with an integer literal, an "
+        "identifier, a blank character or a non-DIL string.\n",
+        input, input);
+}
+
+/*
+This function is used for testing various input strings as arguments of the
+identify_token function.
+*/
+int main(void) {
+  init_isIdChar();
+
+  test("", TOK::EOI);
+  test("int i;", TOK::INT);
+  test("bool var;", TOK::BOOL);
+  test("struct Student;", TOK::STRUCT);
+  test("new int[5];", TOK::NEW);
+  test("func swap(arr : int[], i, j : int)", TOK::FUNC);
+  test("if (x>10)", TOK::IF);
+  test("else{", TOK::ELSE);
+  test("return x;", TOK::RETURN);
+  test("while (i<10)", TOK::WHILE);
+  test("false || true", TOK::FALSE);
+  test("true||false", TOK::TRUE);
+
+  test("01210 + 100", TOK::UNDEFINED);
+  test("var_name", TOK::UNDEFINED);
+  test("intvar", TOK::UNDEFINED);
+  test("function", TOK::UNDEFINED);
+  test(" int x;", TOK::UNDEFINED);
+
```

Comment:
If the first character is a blank character then kind is set to TOK::UNDEFINED.
This is something I expected from my implementation. Is it ok with what you had in mind for this function?

Reply:
No. For one, the lexer eventually will never put the `UNDEFINED` kind in `token`. If it reaches the point that it would `UNDEFINED`, then it means we have a fatal error and we exit the program.
But for now, think that your function should put the next _available_ kind in `token`. Since there is an available token, it should not put `UNDEFINED` (and of course that means skipping whitespace).